### PR TITLE
Remove consolelink object when CheCluster is removed

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
Remove consolelink object when CheCluster is removed

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/18501

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse/che-operator/blob/master/README.md#updating-custom-resource-definition-file)
- [ ] [Nightly OLM bundle is up to date](https://github.com/eclipse/che-operator/blob/master/README.md#update-nightly-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/consolelink.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	_, err := Delete(deployContext, client.ObjectKey{Name: DefaultConsoleLinkName()}, &consolev1.ConsoleLink{})
+	if err != nil {
+		// failed to delete, shouldn't us prevent from removing finalizer
+		logrus.Error(err)
+	}
+
+	return DeleteFinalizer(deployContext, ConsoleLinkFinalizerName)
+}
+
+func hasConsolelinkObject(deployContext *DeployContext) bool {
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/consolelink.go:119]: hasConsolelinkObject -> hasConsoleLinkObject

File: pkg/deploy/consolelink_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/consolelink_test.go:2]:
```suggestion
// Copyright (c) 2012-2021 Red Hat, Inc.
```

File: pkg/deploy/finalizer.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/finalizer.go:2]:
```suggestion
// Copyright (c) 2012-2021 Red Hat, Inc.
```

File: pkg/deploy/sync.go [hunk 4]
```diff
@@ -9,157 +9,236 @@ import (
...
+	logrus.Infof("Deleting object: %s, name: %s", kind, key.Name)
+
+	return doDelete(client, actual)
 }
 
+// Updates object.
+// Returns true if object is up to date otherwiser return false
 func Update(deployContext *DeployContext, actual runtime.Object, blueprint metav1.Object, diffOpts cmp.Option) (bool, error) {
-	client := getClientForObject(blueprint, deployContext)
 	actualMeta := actual.(metav1.Object)
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/sync.go:147]: Should we check object casting here?

File: pkg/controller/che/che_controller.go [hunk 5]
```diff
@@ -137,7 +136,7 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		if err := corev1.AddToScheme(mgr.GetScheme()); err != nil {
 			logrus.Errorf("Failed to add OpenShift Core to scheme: %s", err)
 		}
-		if hasConsolelinkObject() {
+		if util.HasAPIResourceName("consolelinks") {
```
Comment (AndrienkoAleksandr) [inline pkg/controller/che/che_controller.go:139]: Maybe console links should be const?

**Review by AndrienkoAleksandr — APPROVED** [review]

**Review by flacatus — APPROVED** [review]

File: pkg/deploy/finalizer.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func DeleteFinalizer(deployContext *DeployContext, finalizer string) error {
+	deployContext.CheCluster.ObjectMeta.Finalizers = util.DoRemoveString(deployContext.CheCluster.ObjectMeta.Finalizers, finalizer)
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/finalizer.go:47]: Maybe we should check if finalizers exists, then remove finalizer and update cheCluster:

if !util.ContainsString(deployContext.CheCluster.ObjectMeta.Finalizers, finalizer) {
.... remove finalizer logic...
}

to make less requests.

File: pkg/deploy/finalizer.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+		err = util.ReloadCheCluster(deployContext.ClusterAPI.Client, deployContext.CheCluster)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func DeleteObjectWithFinalizer(deployContext *DeployContext, key client.ObjectKey, blueprint metav1.Object, finalizer string) error {
+	_, err := Delete(deployContext, key, blueprint)
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/finalizer.go:65]: Maybe we should check if finalizers exists, then remove finalizer and update cheCluster:

if !util.ContainsString(deployContext.CheCluster.ObjectMeta.Finalizers, finalizer) {
.... remove finalizer logic...
}

to make less requests.

File: pkg/deploy/sync.go [hunk 8]
```diff
@@ -9,157 +9,249 @@ import (
...
 // Sync syncs the blueprint to the cluster in a generic (as much as Go allows) manner.
-// Returns true if the object was created or updated, false if there was no change detected.
+// Returns true if object is up to date otherwiser returns false
 func Sync(deployContext *DeployContext, blueprint metav1.Object, diffOpts cmp.Option) (bool, error) {
-	key := client.ObjectKey{Name: blueprint.GetName(), Namespace: blueprint.GetNamespace()}
-
-	actual, err := Get(deployContext, key, blueprint)
-	if err != nil {
-		return false, err
+	if !deployContext.CheCluster.ObjectMeta.DeletionTimestamp.IsZero() {
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/sync.go:20]: Maybe we should apply comment here and explain, that we returns true for fine finalization. Or update doc comment for 'Sync' method and apply this info there.

**Pull request merged**